### PR TITLE
feat: inventory.json support for child devices

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Enable users to specify initial inventory data for child devices running tedge-agent  using an inventory.json file.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/3430

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s. You can activate automatic signing by running `just prepare-dev` once)
- [x] I ran `just format` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `just check` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot [hunk 1]
```diff
@@ -0,0 +1,68 @@
+*** Settings ***
+Resource            ../../../resources/common.resource
+Library             DateTime
+Library             String
+Library             Cumulocity
+Library             ThinEdgeIO
+
+Test Setup          Custom Setup
+Test Teardown       Custom Teardown
```
Comment (Bravo555) [inline tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot:9]:
```suggestion
Suite Setup          Custom Setup
Suite Teardown       Custom Teardown
```

File: crates/core/tedge_agent/src/twin_manager/actor.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            }
+        };
+        let inventory_json: Value = serde_json::from_reader(file).map_err(|err| {
+            RuntimeError::ActorError(
+                format!("Failed to parse inventory file contents as JSON: {}", err).into(),
+            )
+        })?;
+        let Value::Object(twin_map) = inventory_json else {
+            return Err(RuntimeError::ActorError(
+                "Invalid inventory.json format: expected a JSON object".into(),
```
Comment (Bravo555) [inline crates/core/tedge_agent/src/twin_manager/actor.rs]: todo: attach the filename to the error

File: crates/core/tedge_agent/src/twin_manager/actor.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                    Ok(Some(next)) => next,
+                    _ => break,
+                };
+            }
+        }
+
+        let device_id = self.config.device_topic_id.clone();
+        for (key, value) in inventory_map {
+            self.publish_twin_data(&device_id, key.clone(), value.clone())
+                .await;
```
Comment (Bravo555) [inline crates/core/tedge_agent/src/twin_manager/actor.rs]: suggestion: don't have to clone here

```suggestion
            self.publish_twin_data(&device_id, key, value).await;
```

File: crates/core/tedge_agent/src/twin_manager/actor.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        // This is to prevent the MQTT actor from crashing while trying to send a twin message to this actor later
+        // after this actor has finished and closed its message box, but the MQTT actor still holds the sender half of it.
+        while self.messages.recv().await.is_some() {
+            // Do nothing
+        }
+        Ok(())
+    }
```
Comment (Bravo555) [inline crates/core/tedge_agent/src/twin_manager/actor.rs]: thought: So as I understand, basically what we do here is:

1. Read the fragments from the inventory file
2. Wait until the end of the "init phase", during which we remove fragments that are already published (our heuristic for "end of the init phase" being messages not being published for 1s)
3. Publish all the remaining fragments
4. Effectively exit, but we're forced to loop because other actors may send us messages despite us not wanting to process them anymore.

So the actor stays alive, despite not doing anything, which sucks a bit, but as the comment says, it's hard to exit without the MQTT actor possibly receiving failures after we exit.

Funnily enough, this usage pattern here could be easily handled by an in-memory broker: we could subscribe for a bit, listen to some messages, and exit without other actors even being aware of us, because sources and sinks would be decoupled, but the actor runtime works more on a peer-to-peer basis, so that's why it's hard.

Even still, perhaps upstream producers being coupled to their downstream consumers such that the producer (actor A) can't output after the consumer (actor B) exits isn't a nice property. Perhaps the runtime should simply remove the consumer message box from producer's sender list and let the producer proceed, only if A doesn't depend on B in any way. But there's also a pattern where A sends message to B and then waits for B's response, in which case it does depend on it and should know ASAP if B exits (it should then probably also exit with an error).

So we have both request/response and publish/subscribe patterns, in this case publish/susbcribe is a bit awkward.

File: crates/core/tedge_agent/src/twin_manager/actor.rs
Comment (Bravo555) [inline crates/core/tedge_agent/src/twin_manager/actor.rs:1]: todo: https://thin-edge.github.io/thin-edge.io/references/mappers/c8y-mapper/#base-inventory-model will also have to be updated.

**Review by Bravo555 — CHANGES_REQUESTED** [review]
One error message adjustment request and docs update, apart from that, looks good and will approve after these changes.

File: crates/core/tedge_agent/src/twin_manager/actor.rs (hunk 4 above)
Comment (albinsuresh) [reply crates/core/tedge_agent/src/twin_manager/actor.rs]: > Even still, perhaps upstream producers being coupled to their downstream consumers such that the producer (actor A) can't output after the consumer (actor B) exits isn't a nice property.

Completely agree. I am really tempted to change the MQTT actor logic to not fail on send when a channel is closed, but just log and proceed (potentially removing that sender from its `peer_senders` list as well). I just wasn't fully sure if some other part of the code is expecting on the existing behaviour to handle things like shutdown requests. That's why I went with this not-so-pretty workaround. But, I'm open to do that change in a separate PR so that we can discuss the impact.

Comment (Bravo555) [reply crates/core/tedge_agent/src/twin_manager/actor.rs]: I'm happy to accept this workaround. As for changing only the MQTT actor (because indeed actors themselves have a choice to exit or keep going if send fails!), also approve making the modification in a separate PR, I'm curious what the test results would be.

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot (hunk 1 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot:9]: Resolved by [8754d74](https://github.com/thin-edge/thin-edge.io/pull/3755/commits/8754d74929df9149acb72f6d9f37bcd9a8a531c9)

File: crates/core/tedge_agent/src/twin_manager/actor.rs (hunk 2 above)
Comment (albinsuresh) [reply crates/core/tedge_agent/src/twin_manager/actor.rs]: Resolved by [8754d74](https://github.com/thin-edge/thin-edge.io/pull/3755/commits/8754d74929df9149acb72f6d9f37bcd9a8a531c9)

File: crates/core/tedge_agent/src/twin_manager/actor.rs (hunk 3 above)
Comment (albinsuresh) [reply crates/core/tedge_agent/src/twin_manager/actor.rs]: Resolved by [8754d74](https://github.com/thin-edge/thin-edge.io/pull/3755/commits/8754d74929df9149acb72f6d9f37bcd9a8a531c9)

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot (hunk 1 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot:9]: Am I missing something? What is the reasoning from using Test vs Suite setups here? Given that there is only one system test in the suite, there will be zero effect, however I don't get the motivation from switching from Test to Suite.

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+*** Keywords ***
+Custom Setup
+    # Parent
+    ${PARENT_SN}=    Setup    connect=${False}
+    Set Suite Variable    $PARENT_SN
+
+    ${parent_ip}=    Get IP Address
+    Set Suite Variable    $PARENT_IP    ${parent_ip}
+    Execute Command    sudo tedge config set c8y.enable.log_upload true
```
Comment (rina23q) [inline tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot]: I think this is unnecessary.

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_bootstrap.robot [hunk 6]
```diff
@@ -6,7 +6,7 @@ Library             ThinEdgeIO
 Test Setup          Custom Setup
 Test Teardown       Get Logs
 
-Test Tags           theme:c8y    theme:telemetry
+Test Tags           theme:c8y
```
Comment (rina23q) [inline tests/RobotFramework/tests/cumulocity/inventory/inventory_update_bootstrap.robot]: You can also remove `theme:telemetry` in `inventory_update.robot`.

Comment (reubenmiller) [inline tests/RobotFramework/tests/cumulocity/inventory/inventory_update_bootstrap.robot]: The `theme:telemetry` tag is still ok here as we're talking about digital twin data (which is a kind of telemetry data)

Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/inventory/inventory_update_bootstrap.robot]: I would actually keep the `theme:telemetry` as digital twin properties is telemetry data.

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot (hunk 1 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot:9]: Closing the comment as generally the users can change the Test/Suite setup to their liking, however a Test Setup/Teardown would be safer than a suite-wide setup as it offers test isolation (with the downside of taking longer to execute).

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_bootstrap.robot (hunk 6 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/cumulocity/inventory/inventory_update_bootstrap.robot]: Resolved in [4612f52](https://github.com/thin-edge/thin-edge.io/pull/3755/commits/4612f524e50ea96e99778d3451efe981b66e8303)

Comment (albinsuresh) [reply tests/RobotFramework/tests/cumulocity/inventory/inventory_update_bootstrap.robot]: Resolved in [4612f52](https://github.com/thin-edge/thin-edge.io/pull/3755/commits/4612f524e50ea96e99778d3451efe981b66e8303)

File: tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot (hunk 5 above)
Comment (albinsuresh) [reply tests/RobotFramework/tests/cumulocity/inventory/inventory_update_child_device.robot]: Resolved in [4612f52](https://github.com/thin-edge/thin-edge.io/pull/3755/commits/4612f524e50ea96e99778d3451efe981b66e8303)

File: crates/core/tedge_agent/src/twin_manager/actor.rs
Comment (albinsuresh) [reply crates/core/tedge_agent/src/twin_manager/actor.rs:1]: Resolved in [4612f52](https://github.com/thin-edge/thin-edge.io/pull/3755/commits/4612f524e50ea96e99778d3451efe981b66e8303)
